### ISSUE TITLE
how to visualize the high response areas?

### ISSUE THREAD

**Issue by erjiaxiao** [dvlab-research/LLaMA-VID#26]
Thanks for such a nice work! Recently I found that it mentions visualization of the high response areas on image when inference. But it seems that the related code is not included in the repo.

**Comment by yanwei-li** [comment]
Hi, we use the feature of `ctx_embed` after [this line](https://github.com/dvlab-research/LLaMA-VID/blob/5a1077586219ed16b31af0f02a7b461881d2ff9f/llamavid/model/llamavid_arch.py#L377) and sort the value to keep the top 20 values. Here is the sample code, please adjust the image according to your setting.
```python
img_score = ctx_embed
img_mean = self.model.get_vision_tower().image_processor.image_mean
img_std = self.model.get_vision_tower().image_processor.image_std
img_show = images[img_idx].permute(1, 2, 0).detach().cpu().numpy()[...,::-1] # set it to raw image
img_show = (img_show * np.array(img_std)[None, None] + np.array(img_mean)[None, None]) * 255
img_score = img_score - img_score.min()
img_score = img_score / max(1, img_score.max())
score_select = img_score.sort(0, descending=True)[0][20]
img_score[img_score<score_select] = 0

img_score = img_score.reshape(int(math.sqrt(len(img_score))), -1)
img_score = nn.functional.interpolate(img_score[None,None], size=(img_show.shape[0], img_show.shape[1]), 
                                    mode='nearest').cpu().numpy() * 255
img_show[...,-1] = img_show[...,-1] + img_score[0][0] * 0.7
```

**Comment by erjiaxiao** [comment]
@yanwei-li thanks a lot! : )

**Comment by erjiaxiao** [comment]
Hi @yanwei-li , I ran the code above and found that:

`
img_score = ctx_embed
img_mean = self.model.get_vision_tower().image_processor.image_mean
img_std = self.model.get_vision_tower().image_processor.image_std
img_show = images[img_idx].permute(1, 2, 0).detach().cpu().numpy()[...,::-1] # set it to raw image
img_show = (img_show * np.array(img_std)[None, None] + np.array(img_mean)[None, None]) * 255
img_score = img_score - img_score.min()
img_score = img_score / max(1, img_score.max())

# This line may cause an error since the dimension of img_score here is (1,32,576). Do you think we should flatten it first?
score_select = img_score.sort(0, descending=True)[0][20]
img_score[img_score<score_select] = 0

img_score = img_score.reshape(int(math.sqrt(len(img_score))), -1)
img_score = nn.functional.interpolate(img_score[None,None], size=(img_show.shape[0], img_show.shape[1]),mode='nearest').cpu().numpy() * 255
img_show[...,-1] = img_show[...,-1] + img_score[0][0] * 0.7`

**Comment by zhangbaijin** [comment]
@erjiaxiao  Did you solved the problem? we should flatten it first?

**Comment by erjiaxiao** [comment]
Hello @zhangbaijin! You could take the mean of the matrix across its second dimension, reducing its shape from (1, 32, 576) to (1, 576). Subsequently, after obtaining the context embedding's image score, you can interpolate it to match the input image's dimensions.

**Comment by zhangbaijin** [comment]
@erjiaxiao Thanks for your answer, my code seem doesn't work. If it is convenient for you, could you please send the code to my email, SemiZxf@sjtu.edu.cn? Thank you very much.

**Comment by erjiaxiao** [comment]
Hi @zhangbaijin, since that period of time I have worked on other projects and I could not find the code I wrote before, but I remember the code provided in this issue could visualize the high response areas.

**Comment by erjiaxiao** [comment]
hi @Jeff-i-Programme, here is my email address: erjiaxiao@gmail.com

**Comment by Jeff-i-Programme** [comment]
> Hi @yanwei-li , I ran the code above and found that a error occurs:
> 
> ```python
> img_score = ctx_embed
> img_mean = self.model.get_vision_tower().image_processor.image_mean
> img_std = self.model.get_vision_tower().image_processor.image_std
> img_show = images[img_idx].permute(1, 2, 0).detach().cpu().numpy()[...,::-1] # set it to raw image
> img_show = (img_show * np.array(img_std)[None, None] + np.array(img_mean)[None, None]) * 255
> img_score = img_score - img_score.min()
> img_score = img_score / max(1, img_score.max())
> 
> # This line may cause an error since the dimension of img_score here is (1,32,576). Do you think we should flatten it first?
> score_select = img_score.sort(0, descending=True)[0][20]
> img_score[img_score<score_select] = 0
> 
> img_score = img_score.reshape(int(math.sqrt(len(img_score))), -1)
> img_score = nn.functional.interpolate(img_score[None,None], size=(img_show.shape[0], img_show.shape[1]),mode='nearest').cpu().numpy() * 255
> img_show[...,-1] = img_show[...,-1] + img_score[0][0] * 0.7
> ```

Hello, I would like to discuss this matter with you. May I have your contact information? Here is my email address [2674729267@qq.com](mailto:2674729267@qq.com)
